host/src/FunctionsNetHost/Grpc/GrpcClient.cs

```diff
@@ -0,0 +1,134 @@
+        {
+            while (await responseStream.MoveNext())
+            {
+                await _messageHandler.ProcessMessageAsync(responseStream.Current);
+            }
+        }
+
+        private async Task StartWriterAsync(IClientStreamWriter<StreamingMessage> requestStream)
+        {
+            await foreach (var rpcWriteMsg in _outgoingMessageChannel.Reader.ReadAllAsync())
+            {
+                await requestStream.WriteAsync(rpcWriteMsg);
+            }
+        }
+
+        private async Task SendStartStreamMessageAsync(IClientStreamWriter<StreamingMessage> requestStream)
+        {
+            var startStreamMsg = new StartStream()
+            {
+                WorkerId = _grpcWorkerStartupOptions.WorkerId
+            };
+
+            var startStream = new StreamingMessage()
+            {
+                StartStream = startStreamMsg
+            };
+
+            await requestStream.WriteAsync(startStream);
+        }
+
+        private FunctionRpcClient CreateFunctionRpcClient(string endpoint)
+        {
+            if (!Uri.TryCreate(endpoint, UriKind.Absolute, out var grpcUri))
+            {
+                throw new InvalidOperationException($"The gRPC channel URI '{endpoint}' could not be parsed.");
+            }
+
+            var grpcChannel = GrpcChannel.ForAddress(grpcUri, new GrpcChannelOptions()
+            {
+                MaxReceiveMessageSize = _grpcWorkerStartupOptions.GrpcMaxMessageLength,
+                MaxSendMessageSize = _grpcWorkerStartupOptions.GrpcMaxMessageLength,
+                Credentials = ChannelCredentials.Insecure
+            });
+
+            return new FunctionRpcClient(grpcChannel);
+        }
+
+        /// <summary>
+        /// Listens to messages in the inbound message channel and forward them the customer payload via interop layer.
+        /// </summary>
+        private async Task StartInboundMessageForwarding()
+        {
+            await foreach (var inboundMessage in MessageChannel.Instance.InboundChannel.Reader.ReadAllAsync())
+            {
+                await HandleIncomingMessage(inboundMessage);
+            }
+        }
+
+        /// <summary>
+        /// Listens to messages in the inbound message channel and forward them the customer payload via interop layer.
+        /// </summary>
+        private async Task StartOutboundMessageForwarding()
+        {
+            await foreach (var outboundMessage in MessageChannel.Instance.OutboundChannel.Reader.ReadAllAsync())
+            {
+                await _outgoingMessageChannel.Writer.WriteAsync(outboundMessage);
+            }
+        }
+
+        private static Task HandleIncomingMessage(StreamingMessage inboundMessage)
+        {
+            // Queue the work to another thread.
+            Task.Run(() =>
+            {
+                byte[] inboundMessageBytes = inboundMessage.ToByteArray();
```

Comment:
Did we review this after switching to managed? Will this copy the bytes over? Wasn't there a span that exposed the message bytes?

Reply:
Hm, the proto compiler generated C# class with the StreamingMessage type, not with ByteBuffer. The generated code does not have any "ToSpan"_ish_ methods. We can explore more on this to see what other optimization we can do here. Perhaps as a separate work item?

Reply:
Opened #1711 to explore this further.